### PR TITLE
feat: Support lora in dtensor grpo workflow[1/3]: sync and colocated setup

### PR DESCRIPTION
**Pull request by RayenTian**
# What does this PR do ?

Add LoRA Support for GRPO with Synchronous Colocated Inference

## Core Features
- LoRA Integration with DTensor Backend: Added support for LoRA configuration in the DTensor policy worker, enabling LoRA-based training with colocated inference
- Selective Weight Refitting: Implemented granular control over weight updates with `refit_base_model_weights` and `refit_lora_weights` parameters in the `refit_policy_generation()` function
- LoRA Weight Streaming: Extended the IPC ZMQ weight streaming mechanism to support both base model weights and LoRA adapter weights

## New Components

`nemo_rl/models/generation/lora.py`: New module containing:
- LoRARequestWithCfgAndWeights: Custom LoRA request class that supports passing configuration and weights directly
- patched_load_adapter(): Patched adapter loading from memory

## Testing
- Functional Test: Added `tests/functional/grpo_automodel_lora.sh` to validate end-to-end LoRA GRPO workflow
- Unit Tests: Extended test_vllm_generation.py with LoRA-specific test cases
- Nightly Test: Added `examples/configs/recipes/llm/grpo-qwen3-8B-base-1n8g-fsdp2-lora.yaml`

# Issues
[1/3] of #1597 

## Subsequent PRs

- [PR! 1751 (feat: Support lora in dtensor grpo workflow[2/3]: sync and non-colocated setup)](https://github.com/NVIDIA-NeMo/RL/pull/1751) 
- [PR! 1752 (feat: Support lora in dtensor grpo workflow[3/3]: async vllm)](https://github.com/NVIDIA-NeMo/RL/pull/1752)

# Usage
* **You can potentially add a usage example below**

```sh
bash tests/functional/grpo_automodel_lora.sh
```

# Result
## Models to Test

- [x] Qwen3-0.6
- [x] Llama-3.2-3B-Instruct
- [x] Qwen2.5-7B-Instruct
- [x] Llama-3.1-8B
- [ ] nvidia/NVIDIA-Nemotron-3-Nano-30B-A3B-Base-BF16

### Qwen/Qwen3-0.6B
image

### Llama-3.2-3B-Instruct
image
image

### Llama-3.1-8B
image
image

### Qwen2.5-7B
image
image

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](/NVIDIA-NeMo/RL/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you run the unit tests and functional tests locally? Visit our [Testing Guide](/NVIDIA-NeMo/RL/blob/main/docs/testing.md) for how to run tests
- [ ] Did you add or update any necessary documentation? Visit our [Document Development Guide](/NVIDIA-NeMo/RL/blob/main/docs/documentation.md) for how to write, build and test the docs.

# Additional Information
* ...

## Summary by CodeRabbit

* **New Features**
  * Added LoRA (Low-Rank Adaptation) fine-tuning support for policy models in GRPO training workflows
  * Introduced configurable LoRA hyperparameters (dimension, alpha, dropout, initialization)
  * Added selective weight update controls to optimize training efficiency

* **New Examples**
  * Added Qwen3-8B base model configuration with LoRA fine-tuning enabled

* **Tests**
  * Added functional and integration test suites for LoRA-based GRPO training

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: nemo_rl/models/generation/interfaces.py [hunk 1]
```diff
@@ -245,7 +245,9 @@ def prepare_refit_info(self, state_dict_info: dict[str, Any]) -> None:
         """Prepare the info for refit."""
         raise NotImplementedError
 
-    def update_weights_via_ipc_zmq(self) -> list[ray.ObjectRef]:
+    def update_weights_via_ipc_zmq(
+        self, refit_base_model_weights: bool = True, refit_lora_weights: bool = False
```
Comment (RayenTian) [inline nemo_rl/models/generation/interfaces.py]: TODO: add lora_cfg

File: nemo_rl/models/generation/vllm/vllm_worker.py [hunk 2]
```diff
@@ -789,6 +813,47 @@ def update_weights_from_collective(self) -> bool:
...
+
+        def _get_model_state_dict(self):
+            model = self.get_model()
+            if model is None:
+                return {}
+            return model.state_dict()
+
+        results = self.llm.collective_rpc(_get_model_state_dict)
+        return results
+
```
Comment (RayenTian) [inline nemo_rl/models/generation/vllm/vllm_worker.py]: TODO: debug data. can be deleted.

File: nemo_rl/algorithms/grpo.py [hunk 3]
```diff
@@ -946,9 +961,15 @@ def refit_policy_generation(
                 )
 
             futures_train = policy.stream_weights_via_ipc_zmq(
-                buffer_size_bytes=buffer_size_bytes, kv_scales=kv_scales
+                buffer_size_bytes=buffer_size_bytes,
+                kv_scales=kv_scales,
+                refit_base_model_weights=refit_base_model_weights,
```
Comment (yuki-97) [inline nemo_rl/algorithms/grpo.py]: feels better to make `refit_base_model_weights` and `refit_lora_weights` to one param `refit_mode` in `stream_weights_via_ipc_zmq` function, since in this function we can only set either them to True.

File: nemo_rl/algorithms/grpo.py [hunk 4]
```diff
@@ -971,6 +992,31 @@ def refit_policy_generation(
...
+            update_success = _perform_refit_weights(
+                refit_base_model_weights=True, refit_lora_weights=False
+            )
+        if refit_lora_weights:
+            update_success = (
+                _perform_refit_weights(
+                    refit_base_model_weights=False, refit_lora_weights=True
+                )
+                and update_success
+            )
```
Comment (yuki-97) [inline nemo_rl/algorithms/grpo.py]: if `refit_base_model_weights is True and refit_lora_weights is False`, which is the typical case of lora, `update_success` will always be False finally.

but `update_success` is useless here since we'll raise in `_perform_refit_weights` if it's False, so we can just remove it.
```suggestion
        if refit_base_model_weights:
            _perform_refit_weights(refit_base_model_weights=True, refit_lora_weights=False)
        if refit_lora_weights:
            _perform_refit_weights(refit_base_model_weights=False, refit_lora_weights=True)
```

File: nemo_rl/algorithms/grpo.py [hunk 5]
```diff
@@ -1013,6 +1059,8 @@ def grpo_train(
         policy_generation = policy  # type: ignore
         NEED_REFIT = False
     POLICY_GENERATION_STALE = True  # tracks if generation needs a refit before running
+    REFIT_BASE_MODEL_WEIGHTS = True
+    REFIT_LORA_WEIGHTS = policy.lora_enabled
```
Comment (yuki-97) [inline nemo_rl/algorithms/grpo.py]: let's follow coderabbit for `policy.lora_enabled`.
```suggestion
    REFIT_LORA_WEIGHTS = getattr(policy, "lora_enabled", False)
```

File: nemo_rl/models/generation/vllm/lora.py [hunk 6]
```diff
@@ -0,0 +1,175 @@
+# Copyright (c) 2025, NVIDIA CORPORATION.  All rights reserved.
```
Comment (yuki-97) [inline nemo_rl/models/generation/vllm/lora.py:1]: let's move this file to under `nemo_rl/models/generation/vllm/` to make it clear. maybe as you know we are integrating sglang.

File: nemo_rl/algorithms/grpo.py (hunk 5 above)
Comment (RayenTian) [reply nemo_rl/algorithms/grpo.py]: Done. Remove both `REFIT_LORA_WEIGHTS` and `REFIT_BASE_MODEL_WEIGHTS`. All judgments are made within refit_policy_generation.

File: nemo_rl/algorithms/grpo.py (hunk 3 above)
Comment (RayenTian) [reply nemo_rl/algorithms/grpo.py]: Done.

File: examples/configs/recipes/llm/grpo-qwen3-8B-base-1n8g-fsdp2-lora.yaml [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+checkpointing:
+  checkpoint_dir: results/grpo-qwen3-8B-base-1n8g-fsdp2-lora
+policy:
+  model_name: Qwen/Qwen3-8B-Base
+  max_total_sequence_length: 4096
+  dtensor_cfg:
+    activation_checkpointing: true
+    lora_cfg:
+      enabled: True
+      dim: 256
```
Comment (yuki-97) [inline examples/configs/recipes/llm/grpo-qwen3-8B-base-1n8g-fsdp2-lora.yaml]: just curious why dim and alpha seems larger than that in sft? is it b/c there's some setting we are comparing with use this setting?

File: examples/configs/recipes/llm/grpo-qwen3-8B-base-1n8g-fsdp2-lora.yaml [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+      alpha: 512
+  sequence_packing:
+    enabled: false
+data:
+  shuffle: false
```
Comment (yuki-97) [inline examples/configs/recipes/llm/grpo-qwen3-8B-base-1n8g-fsdp2-lora.yaml]: why here is false? is it a debug thing that forget to remove? 
```suggestion
  shuffle: true
```

File: nemo_rl/algorithms/grpo.py [hunk 9]
```diff
@@ -460,6 +460,19 @@ def setup(
         )
         policy_config["megatron_cfg"]["train_iters"] = total_train_iters
 
+    if policy_config.get("dtensor_cfg", {}).get("enabled", False):
```
Comment (yuki-97) [inline nemo_rl/algorithms/grpo.py]: nit: how about this? also the below.
```suggestion
    if "dtensor_cfg" in policy_config and policy_config["dtensor_cfg"]["enabled"]:
```

File: nemo_rl/algorithms/grpo.py [hunk 10]
```diff
@@ -460,6 +460,19 @@ def setup(
...
+        lora_cfg = policy_config.get("dtensor_cfg", {}).get("lora_cfg", {})
+        if lora_cfg.get("enabled", False):
+            # Override the vLLM lora config with the DTensor lora config
+            generation_config["vllm_cfg"]["lora_cfg"] = lora_cfg
```
Comment (yuki-97) [inline nemo_rl/algorithms/grpo.py:475]: it should be okay for this PR, but do you have any ideas how to unify it when supporting grpo lora in mcore?
also cc @vadam5

File: nemo_rl/models/generation/vllm/lora.py [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+    )
+
+    # patch the load_adapter function in LRUCacheWorkerLoRAManager
+    from vllm.lora.worker_manager import LRUCacheWorkerLoRAManager
+
+    setattr(LRUCacheWorkerLoRAManager, "_load_adapter", patched_load_adapter)
+
+
+# Please Review: Not sure put it here or in nemo_rl/models/generation/vllm/utils.py
+def get_vllm_lora_metadata() -> dict[str, Any]:
```
Comment (yuki-97) [inline nemo_rl/models/generation/vllm/lora.py]: I think just put it here is fine since it's only related to lora.

File: nemo_rl/models/generation/vllm/vllm_backend.py [hunk 12]
```diff
@@ -125,8 +125,122 @@ def _maybe_process_fp8_kv_cache(self) -> None:
             target_device,
         )
 
+    def _apply_weight_name_mapping(
```
Comment (yuki-97) [inline nemo_rl/models/generation/vllm/vllm_backend.py]: seems this function is only related to lora, can we move it to `nemo_rl/models/generation/vllm/lora.py` to make here clear?

File: nemo_rl/models/generation/vllm/vllm_backend.py [hunk 13]
```diff
@@ -125,8 +125,122 @@ def _maybe_process_fp8_kv_cache(self) -> None:
...
+        new_weights = []
+        for name, w in weights:
+            new_name = map_param_name(name)
+            new_weights.append((new_name, w))
+        return new_weights
+
+    def _apply_loaded_weights(
+        self,
+        weights: list[tuple[str, torch.Tensor]],
+        lora_config: dict[str, Any],
```
Comment (yuki-97) [inline nemo_rl/models/generation/vllm/vllm_backend.py]: will `lora_config` change in different request?

if not, feels better to do a `prepare_lora_info` like `prepare_refit_info`. we can record these things there and reuse them instead of passing `lora_config` each time.
1. `lora_enabled`
2. `lora_cfg_dict`
3. `lora_metadata`

File: nemo_rl/models/generation/vllm/vllm_backend.py [hunk 14]
```diff
@@ -125,8 +125,122 @@ def _maybe_process_fp8_kv_cache(self) -> None:
...
+        assert refit_mode in ["base_model", "lora"], (
+            f"refit_mode must be 'base_model' or 'lora', but got {refit_mode}"
+        )
+
+        from nemo_rl.models.generation.vllm.quantization import fp8
+
+        runner = self.model_runner
+
+        if fp8.is_fp8_model(runner.vllm_config):
+            # the fp8 load_weights additionally casts bf16 weights into fp8
```
Comment (yuki-97) [inline nemo_rl/models/generation/vllm/vllm_backend.py:158]: what about adding `assert refit_mode == "base_model"` here?

File: nemo_rl/models/policy/lm_policy.py [hunk 15]
```diff
@@ -885,3 +894,23 @@ def print_node_ip_and_gpu_id(self) -> list[tuple[str, int]]:
...
+
+        if is_megatron or (is_dtensor and not is_v2) or (not lora_enabled):
+            return False, False
+
+        # Check if the base model weights have been refit only when LoRA is enabled and DTensor v2 is used
+        futures = self.worker_group.run_all_workers_single_data(
+            "get_lora_base_refit_done"
+        )
+        results = ray.get(futures)
+        return True, bool(results) and all(results)
```
Comment (yuki-97) [inline nemo_rl/models/policy/lm_policy.py]: just curious, why we need `bool(results)` here?

File: tests/test_suites/llm/grpo-qwen3-8B-base-1n8g-fsdp2-lora.sh [hunk 16]
```diff
@@ -0,0 +1,41 @@
...
+uv run tests/json_dump_tb_logs.py $LOG_DIR --output_path $JSON_METRICS
+
+# Only run metrics if the target step is reached
+if [[ $(jq 'to_entries | .[] | select(.key == "train/loss") | .value | keys | map(tonumber) | max' $JSON_METRICS) -ge $MAX_STEPS ]]; then
+    uv run tests/check_metrics.py $JSON_METRICS \
+        'mean(data["train/gen_kl_error"]) < 0.001' \
+        'data["train/gen_kl_error"]["20"] < 0.001' \
+        'mean(data["train/reward"]) > 0.56' \
+        'mean(data["timing/train/total_step_time"], 2) < 50'
+fi
```
Comment (yuki-97) [inline tests/test_suites/llm/grpo-qwen3-8B-base-1n8g-fsdp2-lora.sh:43]:
```suggestion
    uv run tests/check_metrics.py $JSON_METRICS \
        'mean(data["train/gen_kl_error"]) < 0.001' \
        'data["train/gen_kl_error"]["20"] < 0.001' \
        'mean(data["train/reward"]) > 0.56' \
        'mean(data["timing/train/total_step_time"], 2) < 50'

    # Clean up checkpoint directory after successful run to save space.
    rm -rf "$CKPT_DIR"
fi
```

**Review by yuki-97 — COMMENTED** [review]
review finished, thanks again for supporting this!

File: examples/configs/recipes/llm/grpo-qwen3-8B-base-1n8g-fsdp2-lora.yaml (hunk 8 above)
Comment (RayenTian) [reply examples/configs/recipes/llm/grpo-qwen3-8B-base-1n8g-fsdp2-lora.yaml]: Just for debug. Now remove it.

File: nemo_rl/models/generation/vllm/vllm_backend.py (hunk 12 above)
Comment (RayenTian) [reply nemo_rl/models/generation/vllm/vllm_backend.py]: Done.

File: nemo_rl/models/generation/vllm/vllm_backend.py (hunk 14 above)
Comment (RayenTian) [reply nemo_rl/models/generation/vllm/vllm_backend.py]: Done.

File: nemo_rl/models/generation/vllm/vllm_backend.py [hunk 17]
```diff
@@ -125,8 +125,51 @@ def _maybe_process_fp8_kv_cache(self) -> None:
...
+        return new_weights
+
     @wrap_with_nvtx_name("vllm_internal_worker_extension/update_weights_via_ipc_zmq")
-    def update_weights_via_ipc_zmq(self) -> bool:
+    def update_weights_via_ipc_zmq(
+        self,
+        lora_config: dict[str, Any] = {},
+        refit_base_model_weights: bool = True,
+        refit_lora_weights: bool = False,
+    ) -> bool:
```
[the remark this answers is not in the document]
Comment (RayenTian) [reply nemo_rl/models/generation/vllm/vllm_backend.py]: Remove the config.
